src/main/java/com/rabbitmq/client/impl/AMQChannel.java

```diff
@@ -153,8 +161,19 @@ public void handleCompleteInboundCommand(AMQCommand command) throws IOException
         // waiting RPC continuation.
         if (!processAsync(command)) {
             // The filter decided not to handle/consume the command,
-            // so it must be some reply to an earlier RPC.
-            RpcContinuation nextOutstandingRpc = nextOutstandingRpc();
+            // so it must be a response to an earlier RPC.
+            if (_checkRpcResponseType) {
+                synchronized (_channelMutex) {
+                    // check if this reply command is intended for the current waiting request before calling nextOutstandingRpc()
+                    if (!_activeRpc.canHandleReply(command)) {
+                        // this reply command is not intended for the current waiting request
+                        // most likely a previous request timed out and this command is the reply for that.
+                        // Throw this reply command away so we don't stop the current request from waiting for its reply
+                        return;
+                    }
+                }
+            }
+            final RpcContinuation nextOutstandingRpc = nextOutstandingRpc();
```

Comment:
its a super small edge case something could change... but to be safe should probably call nextOutstandingRpc under original channelMutex lock instead of releasing and reaquiring it.